### DISCUSSION TITLE
How to remove decorator for new line

### DISCUSSION THREAD

**Discussion by JCzz** [microsoft/vscode-discussions#907] [Extension Development QnA] [answered]
I have added a decorator to a line, but when hitting enter on the end of the line, the decorator follows down to the new line. The decorator is just an underline:
```

    const terminal = vscode.window.activeTerminal;

    const activeEditor = vscode.window.activeTextEditor;

    let lineText: string;
    if (activeEditor) {
        let lineNumber = activeEditor.selection.active.line;
        lineNumber = lineNumber;
        const line = activeEditor.document.lineAt(lineNumber);
        lineText = activeEditor.document.lineAt(lineNumber).text;

        // decorationType2 = vscode.window.createTextEditorDecorationType({});



        const decoration = {
            range: new vscode.Range(line.range.start, line.range.end),
        };

        activeEditor.setDecorations(deco, [decoration]);
```
Problem is that when hitting enter, the decorator from the line moves down to new line. So I am trying to remove the decorator for this newline:
```
import * as vscode from 'vscode';

export function attachCleanLine(deco: vscode.TextEditorDecorationType) {
	vscode.workspace.onDidChangeTextDocument(event => {
		const activeEditor = vscode.window.activeTextEditor;
		if (!activeEditor) {return;}

		const contentChange = event.contentChanges[0];
		if (contentChange.text === '\n') {
			let currentLine = activeEditor.selection.active.line;
			currentLine = currentLine + 1;
			let lineText = activeEditor.document.lineAt(currentLine).text;
			console.log(`Enter key pressed on line ${currentLine}, someVariable: ${lineText}`);

			const line = activeEditor.document.lineAt(currentLine);


			// HOWTO: remove decorator for the new line after return
			activeEditor.setDecorations(deco, []);
		}
	});
}
```
But it does not get removed, do you know how to remove a decorator from a line?

The repo is here: https://github.com/appsyouwear/linedance

**Comment by tamuratak** [comment, marked as answer]
The following would help:

1. Create a new decoration with  `createTextEditorDecorationType` every time you call `setDecorations`.
2. Remove old unnecessary ones by calling `decorationType.dispose()`: See the [doc](https://code.visualstudio.com/api/references/vscode-api#TextEditorDecorationType).

> Remove this decoration type and all decorations on all text editors using it.

**Reply by JCzz** [reply]
Thanks

**Comment by r3m0t** [comment]
Hi, disposing the text decoration type and recreating it is really inefficient.

I think VS Code is growing the range of the text decoration to include the next line. This happens before your `onDidChangeTextDocument` event so the user sees the decoration on the next line until the event fires.

I think you want to set `rangeBehavior` to prevent the range from growing. Try different values.

As well, consider the user might paste something with a new line, or use `'\r\n'` as the line ending in the file, so use `contentChange.text.includes('\n')` . And, the user might have multiple cursors or make multiple edits using 'Replace All', so use `for (const contentChange of event.contentChanges)` .
